### PR TITLE
Numeric in/equality and comparisons across numeric types

### PR DESCRIPTION
**Pull request by TristonianJones**

### CODE REVIEW

File: common/types/compare.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	}
+	return compareDouble(d, Double(u))
+}
+
+func compareUintDouble(u Uint, d Double) Int {
+	return -compareDoubleUint(d, u)
+}
+
+func compareIntUint(i Int, u Uint) Int {
+	if i < 0 || u > math.MaxInt64 {
```
Comment (jcking) [inline common/types/compare.go:50]: Doesn't this cause `compareIntUint(math.MaxInt64, math.MaxUint64)` to be -1? Is that intentional?

File: common/types/double.go [hunk 2]
```diff
@@ -58,17 +59,19 @@ func (d Double) Add(other ref.Val) ref.Val {
...
-	otherDouble, ok := other.(Double)
-	if !ok {
+	switch ov := other.(type) {
+	case Double:
+		if math.IsNaN(float64(d)) || math.IsNaN(float64(ov)) {
+			return NewErr("NaN values cannot be ordered")
+		}
+		return compareDouble(d, ov)
+	case Int:
+		return compareDoubleInt(d, ov)
```
Comment (jcking) [inline common/types/double.go:72]: Shouldn't we check for d being NaN here as well?

File: common/types/double.go [hunk 3]
```diff
@@ -58,17 +59,19 @@ func (d Double) Add(other ref.Val) ref.Val {
...
+	case Uint:
+		return compareDoubleUint(d, ov)
```
Comment (jcking) [inline common/types/double.go:74]: Same.

**Review by jcking — CHANGES_REQUESTED** [review]

File: common/types/compare.go (hunk 1 above)
Comment (TristonianJones) [reply common/types/compare.go:50]: Yes, that's right. `int64` max is less than `uint64` max.

File: common/types/double.go (hunk 2 above)
Comment (TristonianJones) [reply common/types/double.go:72]: Good point. Done.

File: common/types/double.go (hunk 3 above)
Comment (TristonianJones) [reply common/types/double.go:74]: Done.

File: common/types/compare.go (hunk 1 above)
Comment (jcking) [reply common/types/compare.go:50]: Oh...wow. Ignore me my brain was not working. lol

**Review by jcking — APPROVED** [review]

**Pull request merged**